File: src/python/2023/day3/part2.py
print(sum([
    v[0] * v[1] for v in  # sums all asterisks with > 2
    [
        [sum([g[w, a] for a in range(140) if abs(w - a*140) < 280], []) for w in range(140**2)] for g in # coagulate the dictionaries
            [{
                (r*140 + p, r): [m[2] for m in zip(*l) if m[0] and not (m[1] and m[2] == m[3]) and m[2] != -1 and r*140 + p in m[0]]  # find adjacent numbers
                    for p in range(-280, 280) for r, l in enumerate( # from part 1.
                        list(map(lambda x, y: (x, [-1] + x, y, [-1] + y), 
                            [[[l for l in n if l > 0] for n in zip(*m)] for m in zip(
                                *list(map(lambda x: [
                                        [[False] + q for q in [[False] * len(x[0])] + x], [[False] * len(x[0])] + x, [q[1:] + [False] for q in [[False] * len(x[0])] + x],
                                        [[False] + q for q in x], x, [q[1:] + [False] for q in x],
                                        [[False] + q for q in x[1:] + [[False] * len(x[0])]],  x[1:] + [[False] * len(x[0])], [q[1:] + [False] for q in x[1:] + [[False] * len(x[0])]]],
                                    [[[i * 140 + j if b == "*" else -1 for j, b in enumerate(a)] for i, a in enumerate(open("resources/2023/day3.txt").read().split("\n"))]] # array of asterisk ids, or -1
                                ))[0]
                            )], # array of nearby asterissks + id
                            [[max(a) if a[1] != -1 else -1 for a in zip(*b)] for b in zip(
                                *list(map(lambda x: [[[-1] + q for q in x], x, [q[1:] + [-1] for q in x]],
                                    [[[int("".join(c).strip('.!@#$%^&*()_+-=/')) if c[1] in '0123456789' else -1 for c in zip('.' + l, l, l[1:] + '.')] for l in open("resources/2023/day3.txt").read().split("\n")]]
                                ))[0]
                            )] # array of -1, or the number.
                        ))
                    )
            }] 
    ][0] if len(v) >= 2 
]))

# sum([v[0] * v[1] for v in [[sum([g[w, a] for a in range(140) if abs(w - a*140) < 280], []) for w in range(140**2)] for g in [{(r*140 + p, r): [m[2] for m in zip(*l) if m[0] and not (m[1] and m[2] == m[3]) and m[2] != -1 and r*140 + p in m[0]] for p in range(-280, 280) for r, l in enumerate(list(map(lambda x, y: (x, [-1] + x, y, [-1] + y), [[[l for l in n if l > 0] for n in zip(*m)] for m in zip(*list(map(lambda x: [[[False] + q for q in [[False] * len(x[0])] + x], [[False] * len(x[0])] + x, [q[1:] + [False] for q in [[False] * len(x[0])] + x],[[False] + q for q in x], x, [q[1:] + [False] for q in x], [[False] + q for q in x[1:] + [[False] * len(x[0])]],  x[1:] + [[False] * len(x[0])], [q[1:] + [False] for q in x[1:] + [[False] * len(x[0])]]], [[[i * 140 + j if b == "*" else -1 for j, b in enumerate(a)] for i, a in enumerate(open("resources/2023/day3.txt").read().split("\n"))]]))[0])], [[max(a) if a[1] != -1 else -1 for a in zip(*b)] for b in zip(*list(map(lambda x: [[[-1] + q for q in x], x, [q[1:] + [-1] for q in x]], [[[int("".join(c).strip('.!@#$%^&*()_+-=/')) if c[1] in '0123456789' else -1 for c in zip('.' + l, l, l[1:] + '.')] for l in open("resources/2023/day3.txt").read().split("\n")]]))[0])])))}]][0] if len(v) >= 2])

File: src/pascal/2023/day3/part2.pas
program day3part2;
uses sysutils;

var
file_: text;
curr: string;
sum, game: int32;
state: int32;
number: int32;
valid: boolean;

i, a, b: int32;
line: int32;
c: char;

LINE_LENGTH: int32 = 140;

curr_asterisk: int32;
asterisks: int32;
asterisk_values: array[1..1000] of int32;
asterisk_locations: array[1..1000, 1..2] of int32;

function is_symbol(str: string; ind: int32): boolean;
begin
    is_symbol := (str[ind] = '*')
end;

function find_asterisk(hpos: int32; vpos: int32): int32;
var
    j: int32;

begin

    find_asterisk := -1;

    for j := 1 to asterisks do 
    begin
        if ((asterisk_locations[j, 1] = hpos) and (asterisk_locations[j, 2] = vpos)) then
        begin
            find_asterisk := j;
            break;
        end;
    end;
end;

function get_asterisk_id(hpos: int32; vpos: int32): int32;
var
    fucked_up: int32;

begin

    get_asterisk_id := -1;

    fucked_up := find_asterisk(hpos-1, vpos-1);
    if (fucked_up <> -1) then
        get_asterisk_id := fucked_up;
    fucked_up := find_asterisk(hpos, vpos-1);
    if (fucked_up <> -1) then
        get_asterisk_id := fucked_up;
    fucked_up := find_asterisk(hpos+1, vpos-1);
    if (fucked_up <> -1) then
        get_asterisk_id := fucked_up;
    fucked_up := find_asterisk(hpos-1, vpos);
    if (fucked_up <> -1) then
        get_asterisk_id := fucked_up;
    fucked_up := find_asterisk(hpos+1, vpos);
    if (fucked_up <> -1) then
        get_asterisk_id := fucked_up;
    fucked_up := find_asterisk(hpos-1, vpos+1);
    if (fucked_up <> -1) then
        get_asterisk_id := fucked_up;
    fucked_up := find_asterisk(hpos, vpos+1);
    if (fucked_up <> -1) then
        get_asterisk_id := fucked_up;
    fucked_up := find_asterisk(hpos+1, vpos+1);
    if (fucked_up <> -1) then
        get_asterisk_id := fucked_up;

end;

(* Here the main program block starts *)
begin
    { open inp.txt for reading }
    assign(file_, 'resources/2023/day3.txt');
    reset(file_);

    { initialize sum and count to 0 }
    sum := 0;

    { read the first two lines }

    line := 1;
    asterisks := 0;


    {find the position of all asterisks}

    while not eof(file_) do
    begin

        readln(file_, curr);
    
        for i := 1 to LINE_LENGTH do
        begin
            if is_symbol(curr, i) then
            begin
                asterisk_locations[asterisks + 1][1] := i;
                asterisk_locations[asterisks + 1][2] := line;

                asterisk_values[asterisks + 1] := 0;

                asterisks := asterisks + 1;
            end;
        end;

        line := line + 1;

    end;

    line := 0;

    reset(file_);

    writeln('boobs');

    { main loop }
    while not eof(file_) do 
    begin

        state := 0;
        line := line + 1;

        readln(file_, curr);

        for i := 1 to LINE_LENGTH do
        begin
            c := curr[i];
            write(c);
            case state of
                    {waiting for number}
                    0:
                    begin
                        {number}
                        if (('0' <= c) and (c <= '9')) then
                        begin
                            {add number to number}
                            val(c, number);
                            state := 1;

                            curr_asterisk := get_asterisk_id(i, line);

                        end;
                    end;
                    {reading number and checking true}
                    1:
                    begin
                        if (('0' <= c) and (c <= '9')) then
                        begin

                            writeln('boobs2');

                            {add number to number}
                            val(c, b);
                            number := number * 10 + b;

                            {look for symbol}
                            b := get_asterisk_id(i, line);
                            if (b <> -1) then
                                curr_asterisk := b;

                        end
                        else 
                        begin
                            state := 0;
                            writeln(curr_asterisk);

                            if (curr_asterisk <> -1) then
                            begin
                                if (asterisk_values[curr_asterisk] = 0) then
                                    asterisk_values[curr_asterisk] := number
                                else if (asterisk_values[curr_asterisk] > 0) then
                                begin
                                    sum := sum + asterisk_values[curr_asterisk] * number;
                                    asterisk_values[curr_asterisk] := -1;
                                end;
                            end;

                        end;
                            
                    end;
            end;

            write(state);

        end;

        if (state = 1) then
        begin
            begin

                if (curr_asterisk <> -1) then
                begin
                    if (asterisk_values[curr_asterisk] = 0) then
                        asterisk_values[curr_asterisk] := number
                    else if (asterisk_values[curr_asterisk] > 0) then
                    begin
                        sum := sum + asterisk_values[curr_asterisk] * number;
                        asterisk_values[curr_asterisk] := -1;
                    end;
                end;

            end;
        end;

    end;

    i := get_asterisk_id(114, 126);
    writeln(i);
    writeln(asterisk_locations[i, 1]);
    writeln(asterisk_locations[i, 2]);
    writeln(sum);


end.
